### ISSUE TITLE
predictor with text prompt using CLIP

### ISSUE THREAD

**Issue by Usama3059** [facebookresearch/segment-anything#6]
Hi, 

I have implemented text prompt-controlled segmentation using selective search and CLIP. Can you suggest any additional techniques I can include? I am considering trying CLIP-GradCAM   #4 

https://github.com/Usama3059/SAMtext/blob/main/SAM%20_predictor_with_text-prompt_using_CLIP_v1.ipynb

**Comment by getorca** [comment]
interesting, I ran your notebook and it picked out the passenger window.

**Comment by helblazer811** [comment]
CRIS maybe https://arxiv.org/abs/2111.15174.

I was considering fine tuning a model on RefCOCO. I was kindof surprised they didn't release the text to mask model. I suspect they have followup work coming out soon that does this.

**Comment by maxi-w** [comment]
I build a small demo where first SAM extracts all objects and then CLIP is used to retrieve the best matching ones: https://github.com/maxi-w/CLIP-SAM

Example for "kiwi":
![example-segmented](https://user-images.githubusercontent.com/23198066/230231847-6c4719f5-5310-4021-a6d8-58e6b8d0fe86.png)

**Comment by helblazer811** [comment]
Did you try and merge adjacent segmentation masks into larger objects? I just played so runs with the web demo and it seems like contiguous objects can sometimes be segmented into many parts. It would be with testing something like this on a dataset like RefCOCO.

**Comment by helblazer811** [comment]
I'm wondering what the memory foot print would be to train a prompt encoder based on CLIP embeddings. It said in the FAQ that they used 256 A100 GPUs for 3-5 days lol!

**Comment by maxi-w** [comment]
> Did you try and merge adjacent segmentation masks into larger objects? I just played around with the web demo and it seems like contiguous objects can sometimes be segmented into many parts. It would be with testing something like this on a dataset like RefCOCO.

It seems that sometimes it already segments both the individual parts and the whole thing

**Comment by Usama3059** [comment]
I have been working on segmenting small objects with low-quality image descriptions, such as car blinkers. I have used a sliding window technique to extract image patches, which were then subjected to clip retrieval, CLIP Grad-CAM, then point extraction and then SAM to get the segmentation results.

Text-prompt: car blinkers
 
Sample Image:
![s2](https://user-images.githubusercontent.com/43669817/230241022-8d931fe1-e430-46b8-a63c-99db4a2a4ff1.png)

gradcam:
![s1](https://user-images.githubusercontent.com/43669817/230241446-14031c91-9c5a-4a71-9628-d906c534af3a.png)

notebook: https://github.com/Usama3059/SAMtext/blob/main/SAM__predictor_with_text_prompt_using_CLIP_v2.ipynb

There are still some pending tasks, I need to integrate Grad-CAM with SAM, plan to test this technique on more images to assess its effectiveness across different scenarios. Lastly,batch processing  when dealing with a large number of patches."

**Comment by helblazer811** [comment]
I put together some code for testing the model on RefCOCO. 

https://github.com/helblazer811/RefSAM

**Comment by dennisushi** [comment]
> I'm wondering what the memory foot print would be to train a prompt encoder based on CLIP embeddings. It said in the FAQ that they used 256 A100 GPUs for 3-5 days lol!

I believe this is for the full model. Presumably training only a CLIP-based text prompt encoder to the same latent space as the bboxes/points (in a self-supervised manner, keeping the existing bbox/point encoder fixed) would be much easier because the rest of the model can be fully ignored? Would  definitely be better in the long term than checking all found objects, and reverse engineering with CLIP?

**Comment by helblazer811** [comment]
Agreed it would be way less. I think CLIP alone would not work that well
though. CLIP doesn’t do cross attention between text and image features. It
really struggles to contrast images with different relative spatial
orientations of objects. Most SOTA approaches for Referring Image
Segmentation do some sort of Vision Language feature fusion.

**Comment by stefanjaspers** [comment]
> I build a small demo where first SAM extracts all objects and then CLIP is used to retrieve the best matching ones: https://github.com/maxi-w/CLIP-SAM
> 
> Example for "kiwi": ![example-segmented](https://user-images.githubusercontent.com/23198066/230231847-6c4719f5-5310-4021-a6d8-58e6b8d0fe86.png)

The example code provided seems to work on the vit_h model only, is that correct? I'm getting errors when trying the vit_l model.

**Comment by YangGangZhiQi** [comment]
Hi, guys, could you provide the FQA links for me? I just could not find it.

**Comment by helblazer811** [comment]
https://segment-anything.com/?fbclid=IwAR21Y_KQ-NPW4eU6Wid6Tzp8Z_EuuXzZ7a7O7WhDofJTt-Dd7Lt7qcYym5U

Bottom of this page.

**Comment by YangGangZhiQi** [comment]
@helblazer811 Thank you very much!

**Comment by cheerss** [comment]
> Usama3059
You may wish to reference our new work: https://arxiv.org/abs/2212.09506, CLIP is Also an Efficient Segmenter: A Text-Driven Approach for Weakly Supervised Semantic Segmentation

**Comment by songxinkuan** [comment]
Two examples of clip_vit_B/32 text embedding, which reduced from 512 to 256 with laion400m image embedding and text embedding. The first image with 'dog' as text prompt, and the second with 'cat' as text prompt.
![20230413-102319](https://user-images.githubusercontent.com/31535074/231631605-4db91017-a505-4912-a49d-526ce4617518.png)

![20230413-102325](https://user-images.githubusercontent.com/31535074/231631622-406d8a58-9212-496b-951e-145870434a34.png)

**Comment by Eli-YiLi** [comment]
> I have been working on segmenting small objects(based on patch size value) with low-quality image descriptions, such as car blinkers. I have used a sliding window technique to extract image patches, which were then subjected to clip retrieval, CLIP Grad-CAM, then point extraction and then SAM to get the segmentation results.
> 
> Text-prompt: car blinkers
> 
> Sample Image: ![s3](https://user-images.githubusercontent.com/43669817/230394397-4e05129b-3568-48e8-8e4d-782130796545.png)
> 
> gradcam: ![s1](https://user-images.githubusercontent.com/43669817/230241446-14031c91-9c5a-4a71-9628-d906c534af3a.png)
> 
> notebook: https://github.com/Usama3059/SAMtext/blob/main/SAM__predictor_with_text_prompt_using_CLIP_v2.ipynb
> 
> There are still some pending tasks, I need to integrate Grad-CAM with SAM, plan to test this technique on more images to assess its effectiveness across different scenarios. Lastly,batch processing when dealing with a large number of patches."

Grad-CAM is only able to apply on ResNets after ignore self-attention for the middle layer.
We propose a solution enables ViTs and works fine on large resolution.

Our work can achieve text to mask with SAM:
https://github.com/xmed-lab/CLIP_Surgery

This is our work about CLIP's explainability. It's able to guide SAM to achieve text to mask without manual points. 

Besides, it's very simple without any fine-tuning, using the CLIP model itself only. 

Furthermore, it enhances many open-vocabulary tasks, like segmentation, multi-label classification, multimodal visualization. 

This is the jupyter demo:
https://github.com/xmed-lab/CLIP_Surgery/blob/master/demo.ipynb

![fig4](https://user-images.githubusercontent.com/49093246/231662220-2b91306c-4dc8-42d1-9fdd-3fbbafadeb26.jpg)

**Comment by mydcxiao** [comment]
following

**Comment by jvpassarelli** [comment]
Thanks for sharing. Inpsired me to write a simple example, similarly adding text prompt to give the segmentation classes. [Code here](https://github.com/jvpassarelli/sam-clip-segmentation/tree/main) .
